Core: add schema id to snapshot

yyanyy opened the pull request:
- Followup of #2096 to add schema id to snapshot. 
- Also addressed the remaining comments in #2096 in this commit
- Will added some util methods for easier retrieving of snapshot/schema information in follow up PRs.

wypoon commented:
I have some questions:
When we switch from v1 format to v2, and a new metadata file is written for an existing table, what schemas are written to the `schemas` list? And in the `snapshot-log`, what `schema-id` is written for the previous snapshots? (Is it not written, i.e., is null? or is it 0?)
In general, if we see a schema id of 0, does that ever represent a specific schema, or does that always represent some undetermined schema? Let me elaborate: (1) Will we ever see a `schema-id` of 0 in a metadata file and if so, does that refer to a unique schema? (2) In code, if a have an instance of a schema and its schemaId is 0, what are the semantics of that schemaId?

wypoon commented on core/src/test/java/org/apache/iceberg/TestSchemaID.java:
```diff
@@ -0,0 +1,164 @@
+    assertSameSchemaMap(ImmutableMap.of(0, oldSchema, 1, updatedSchema), table.schemas());
+  }
+
+  private void validateSnapshotsAndHistoryEntries(int numElement) {
```
You could have this helper method take a `List<Integer>` containing the schema ids instead of just an int (number of entries). Then `testSchemaIdChangeInSchemaUpdate` would be able to use this helper method for the cases after updating the schema. `snapshots.size()` should then match the `List`'s size, and the `snapshot.schemaId()` in each snapshot in `snapshots` should match the corresponding `Integer` in the `List`.

yyanyy replied:
Great suggestion, this makes the code in this class much easier to read; thank you!

wypoon commented on core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java:
```diff
@@ -22,13 +22,21 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public class SnapshotUtil {
```
I don't see the static methods added here, `snapshotIdFromTime` and `schemaOfSnapshot`, used anywhere in this change. If that is the case, this file can be excluded from this particular PR.

yyanyy replied:
sounds good, will remove and create a separate PR for that.

wypoon commented on core/src/test/resources/TableMetadataV2Valid.json:
```diff
@@ -81,8 +81,39 @@
+    {
+      "snapshot-id": 3051729675574597004,
+      "timestamp-ms": 1515100955770
+    },
```
This suggests to me that not all snapshots in the `snapshot-log` need to have a `schema-id`. Is this what happens if the snapshot was written when the format was v1?

yyanyy replied:
yes, we couldn't get the past schema/id for old snapshots, so I think this is the time when we need to go back to your PR's change to figure out the correct id.

yyanyy commented:
> I have some questions:
> When we switch from v1 format to v2, and a new metadata file is written for an existing table, what schemas are written to the `schemas` list? And in the `snapshot-log`, what `schema-id` is written for the previous snapshots? (Is it not written, i.e., is null? or is it 0?)
> In general, if we see a schema id of 0, does that ever represent a specific schema, or does that always represent some undetermined schema? Let me elaborate: (1) Will we ever see a `schema-id` of 0 in a metadata file and if so, does that refer to a unique schema? (2) In code, if we have an instance of a schema and its schemaId is 0, what are the semantics of that schemaId?

Thank you for the review, and sorry for the delay responding!

I think this change applies to v1 tables as well. When the engine starts to use a release with this change, the new `schemas` list will be written with the current schema and 0 as its default schema id. And in `snapshot-log`, previous snapshots will have null schema-id since they were not available when they were written. 

0 is a valid schema-id and it will refer to a unique schema in metadata file; if there's no schema evolution after the table starts to write `schemas`, 0 will be assigned to the current schema. And in the code, since we only care about id during the interaction with table metadata, and throughout the process when schema class is used as various classes for doing projection etc, schemaId will always be 0, and that is just a default value and shouldn't be used. #2096 has some conversation around this, and this behavior is mentioned in [schema class](https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/Schema.java#L44-L45).

rdblue commented on core/src/main/java/org/apache/iceberg/TableMetadata.java:
```diff
@@ -127,10 +127,16 @@ static TableMetadata newTableMetadata(Schema schema,
+    private final Integer schemaId;
 
-    SnapshotLogEntry(long timestampMillis, long snapshotId) {
+    SnapshotLogEntry(long timestampMillis, long snapshotId, Integer schemaId) {
```
Why add the schema ID to the snapshot log as well as the snapshot? The snapshots in the log are all available in table metadata, so it doesn't seem like there is a benefit to adding it.

yyanyy replied:
Thank you for the review! I think I originally added it to history so that we can directly query history entries to get both snapshot ID and schema ID when we do time based time travel queries, but later I didn't end up using it for the utility methods I [used to have](https://github.com/apache/iceberg/pull/2275/commits/bc534a8f9551c4fb12226c2121001b7cf98f2bb1) so I think we can drop it. I'll update the PR to reflect this.

wypoon replied:
The title and description of the PR can be amended to reflect this then.

wypoon commented:
@yyanyy thank you for responding to my question. But because I am slow, I needed to do some testing to understand for myself the effects of https://github.com/apache/iceberg/pull/2096 and this PR.
I started with Iceberg 0.11.0 and created some tables, inserting data into them, and altering the schema. Then I added the commit for https://github.com/apache/iceberg/pull/2096 and the changes here on top (actually I added the commit for https://github.com/apache/iceberg/pull/2089 first, so that the backports are clean). I made further updates to my Iceberg tables, adding data and altering the schema.
From what I see, when the table is updated and a new metadata.json file for the table is written, the metadata.json file still has a `schema` field, but the schema now has a `schema-id` field; it also has a `schemas` field containing an array of schemas, and a `current-schema-id` field. The only schema known at the time of the switchover is the current schema of the table, so if the table change is just change to data, the existing schema is given a `schema-id` of 0, and the schemas array contains a single schema; if the table change is a schema change, then the old schema is given a `schema-id` of 0, the new schema is given a `schema-id` of 1, and the `schemas` array contains two schemas (and the `current-schema-id` field is 1). In the `snapshots` field, any snapshot created before the switchover does not have a `schema-id`; snapshots created after the switchover are given a `schema-id`, corresponding to the schema current at the time the snapshot is written.
When `Snapshot#schemaId()` is called, if the snapshot is written before the switchover, null is returned, and if written after the switchover, a non-null Integer is returned.
It is as you wrote.
With this change, I think it is straightforward to update my PR, https://github.com/apache/iceberg/pull/1508. I just need to update the following method I'm adding to `BaseTable`
```
  public Schema schemaForSnapshot(long snapshotId) {
    TableMetadata current = ops.current();
    // First, check if the snapshot has a schema id associated with it
    Snapshot snapshot = current.snapshot(snapshotId);
    Integer schemaId = snapshot.schemaId();
    if (schemaId != null) {
      return current.schemasById().get(schemaId);
    }
    // Otherwise, read each of the previous metadata files until we find one whose current
    // snapshot id is the snapshot id
    ...
  }
```
I rebuilt Iceberg with my change on top of the previous changes, and I was able to see that the correct schema is used when viewing any snapshot (time travel).
When this change is merged, I will update my PR. I hope that it can be considered then.

yyanyy commented:
> @yyanyy thank you for responding to my question. But because I am slow, I needed to do some testing to understand for myself the effects of #2096 and this PR.
> I started with Iceberg 0.11.0 and created some tables, inserting data into them, and altering the schema. Then I added the commit for #2096 and the changes here on top (actually I added the commit for #2089 first, so that the backports are clean). I made further updates to my Iceberg tables, adding data and altering the schema.
> From what I see, when the table is updated and a new metadata.json file for the table is written, the metadata.json file still has a `schema` field, but the schema now has a `schema-id` field; it also has a `schemas` field containing an array of schemas, and a `current-schema-id` field. The only schema known at the time of the switchover is the current schema of the table, so if the table change is just change to data, the existing schema is given a `schema-id` of 0, and the schemas array contains a single schema; if the table change is a schema change, then the old schema is given a `schema-id` of 0, the new schema is given a `schema-id` of 1, and the `schemas` array contains two schemas (and the `current-schema-id` field is 1). In the `snapshots` field, any snapshot created before the switchover does not have a `schema-id`; snapshots created after the switchover are given a `schema-id`, corresponding to the schema current at the time the snapshot is written.
> When `Snapshot#schemaId()` is called, if the snapshot is written before the switchover, null is returned, and if written after the switchover, a non-null Integer is returned.
> It is as you wrote.
> With this change, I think it is straightforward to update my PR, #1508. I just need to update the following method I'm adding to `BaseTable`
> 
> ```
>   public Schema schemaForSnapshot(long snapshotId) {
>     TableMetadata current = ops.current();
>     // First, check if the snapshot has a schema id associated with it
>     Snapshot snapshot = current.snapshot(snapshotId);
>     Integer schemaId = snapshot.schemaId();
>     if (schemaId != null) {
>       return current.schemasById().get(schemaId);
>     }
>     // Otherwise, read each of the previous metadata files until we find one whose current
>     // snapshot id is the snapshot id
>     ...
>   }
> ```
> 
> I rebuilt Iceberg with my change on top of the previous changes, and I was able to see that the correct schema is used when viewing any snapshot (time travel).
> When this change is merged, I will update my PR. I hope that it can be considered then.

Thank you for spending time verifying the changes, and described the steps here!

wypoon commented on api/src/main/java/org/apache/iceberg/Snapshot.java:
```diff
@@ -126,4 +126,13 @@
+   * @return schema id associated with this snapshot
+   */
+  default Integer schemaId() {
+    return null;
```
Yes, schemaId() returns null in the case where the snapshot was written before this change. Note though, that even after this change, new metadata can have snapshots without schema id (so schemaId() for those snapshots will return null), if it is metadata for a table existing before this change.

wypoon replied:
I have a PR (https://github.com/apache/iceberg/pull/1508) that reads previous metadata to get the schema for the snapshot in case `Snapshot#schemaId()` returns null.

yyanyy replied:
Sorry for the delay in response, and thank you @openinx for the review! And thank you @wypoon for responding!

yyanyy commented on core/src/main/java/org/apache/iceberg/SerializableTable.java:
```diff
@@ -147,6 +147,7 @@ public String location() {
     return properties;
   }
 
+  // Note that schema parsed from string does not contain the correct schema ID.
```
This is related to the comment I had in https://github.com/apache/iceberg/pull/2465#discussion_r612661016 that we didn't persist schema id within `toJson()` which is called when we serialize the table. 

Honestly for this case I think we can go either way; we can use a different `toJson` implementation when serialize the table since schema at that time almost guaranteed to be the original schema from table metadata; however since the only usage of `schemaId` is for time travel queries, and this use case doesn't need id from the current `schema` itself so adding it isn't necessary, and not having schema Id is not something we don't expect per [this note](https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/Schema.java#L44-L45).

rdblue replied:
I think we should use the correct `toJson` implementation so that we don't need to note that the ID doesn't match.

yyanyy replied:
This is actually before we decided to write down schema id as part of `toJson`; now that `toJson` always writes schema id, this note is no longer relevant and I'll remove it.

rdblue commented on core/src/main/java/org/apache/iceberg/BaseSnapshot.java:
```diff
@@ -87,7 +91,18 @@
+    this(io, snapshotId, parentId, timestampMillis, operation, summary, null, dataManifests);
+  }
+
+  BaseSnapshot(FileIO io,
```
Do we need to introduce an additional constructor? Since these constructors are package-private, we can remove them if we can refactor all of the uses.

yyanyy replied:
This was actually me trying to avoid changing a lot of existing code since there are a lot of usage for the old constructor; but since they are all in tests and I'd expect less merge conflicts from tests, I think it might be fine to directly change. Will update.

rdblue commented on core/src/main/java/org/apache/iceberg/TableMetadataParser.java:
```diff
@@ -210,6 +211,21 @@ private static void toJson(TableMetadata metadata, JsonGenerator generator) thro
+    generator.writeEndObject();
+  }
+
+  private static void writeSnapshotRelated(TableMetadata metadata, JsonGenerator generator) throws IOException {
```
It doesn't look like this refactor is necessary any more. It just creates a method that is only used once. I don't think that this file needs to change at all.

rdblue commented on core/src/test/java/org/apache/iceberg/TestSchemaID.java:
```diff
@@ -0,0 +1,131 @@
+
+  private void validateSingleSchema() {
+    Assert.assertEquals("Current schema ID should match",
+        0, table.schema().schemaId());
```
Nit: Does this need the newline?

rdblue commented on core/src/test/java/org/apache/iceberg/TestSchemaID.java:
```diff
@@ -0,0 +1,131 @@
+    table.newAppend().appendFile(FILE_A).appendFile(FILE_B).commit();
+
+    validateSingleSchema();
+    validateSnapshotsAndHistoryEntries(ImmutableList.of(0));
```
It isn't very obvious what is being validated by these methods. I think it would be better to make them more readable so that the context is obvious. I think I would write it more like this:

```
Map<Integer, Schema> schemaMap(Schema... schemas);

...
assertEquals("Should have only one defined schema", schemaMap(table.schema()), table);
assertEquals("Schema id should be the current", table.schema().schemaId(), table.currentSnapshot().schemaId());
```

I think that what you're doing in `validateSnapshotAndHistoryEntries` does make sense, but it would be more readable if you did more of the work in this method:

```java
int onlyId = table.schema().schemaId();
assertEquals("Schema ids should be correct in snapshots",
    ImmutableList.of(onlyId, onlyId),
    Iterables.transform(table.snapshots(), Snapshot::schemaId);
```

I think the extra work of extracting the schema ids is worth the readability. I can see where the ids in the expected list are coming from, and I can see that we're validating that list against the schema ids of the snapshots. If you wanted to save room, you could make a method (`schemaIds`) to reuse the `Iterables.transform` or `Lists.transform` part.

rdblue commented on core/src/test/java/org/apache/iceberg/TestSchemaID.java:
```diff
@@ -0,0 +1,131 @@
+    // update schema
+    table.updateSchema().addColumn("data2", Types.StringType.get()).commit();
+
+    Schema updatedSchema = new Schema(1,
```
Can't we just get the table's current schema? Or is this intended to check that schema evolution produces the right ID in addition to the snapshot tracking?

yyanyy replied:
Yeah I was hoping to make the check explicitly that schema id is now different

rdblue commented on core/src/test/java/org/apache/iceberg/TestSchemaID.java:
```diff
@@ -0,0 +1,131 @@
+        optional(3, "data2", Types.StringType.get())
+    );
+
+    validateTwoSchemas(updatedSchema, oldSchema);
```
Here, I think it is more readable to use what I suggested above:

```java
assertEquals("Should have two schemas", schemaMap(startSchema, updatedSchema), table);
```

rdblue commented on core/src/test/java/org/apache/iceberg/TestSchemaID.java:
```diff
@@ -0,0 +1,131 @@
+            schemaIds.get(i), snapshots.get(i).schemaId()));
+
+    Assert.assertEquals("Number of history entries should match",
+        schemaIds.size(), table.history().size());
```
Do we need to have history assertions if history is not changing?

rdblue commented on core/src/test/java/org/apache/iceberg/TestSnapshotJson.java:
```diff
@@ -126,5 +146,7 @@ public void testJsonConversionWithManifestList() throws IOException {
     Assert.assertNull("Operation should be null", snapshot.operation());
     Assert.assertNull("Summary should be null", snapshot.summary());
+    Assert.assertEquals("Schema ID should match",
+        expected.schemaId(), snapshot.schemaId());
```
Nit: could this assertion be on one line?

rdblue commented on core/src/test/java/org/apache/iceberg/TestTableMetadata.java:
```diff
@@ -93,12 +93,11 @@ public void testJsonConversion() throws Exception {
 
     List<HistoryEntry> snapshotLog = ImmutableList.<HistoryEntry>builder()
         .add(new SnapshotLogEntry(previousSnapshot.timestampMillis(), previousSnapshot.snapshotId()))
-        .add(new SnapshotLogEntry(currentSnapshot.timestampMillis(), currentSnapshot.snapshotId()))
```
Why is this entry no longer present?

yyanyy replied:
I modified it earlier when I put schema id in snapshot entry, and later took it out by mistake as I thought this line was introduced by me. Will make sure to double check next time!

rdblue commented on core/src/test/java/org/apache/iceberg/TableTestBase.java:
```diff
@@ -308,6 +308,8 @@ void validateSnapshot(Snapshot old, Snapshot snap, Long sequenceNumber, DataFile
 
     Assert.assertFalse("Should find all files in the manifest", newPaths.hasNext());
+
+    Assert.assertEquals("Schema ID should match", Integer.valueOf(table.schema().schemaId()), snap.schemaId());
```
Nit: should this just cast to `Integer` like the test case below?

stevenzwu commented on api/src/test/java/org/apache/iceberg/TestHelpers.java:
```diff
@@ -120,6 +120,22 @@ public static void assertSerializedAndLoadedMetadata(Table expected, Table actua
+
+      Assert.assertEquals("Should have matching schema id",
+          schema1.schemaId(), schema2.schemaId());
+      Assert.assertEquals("Should have matching schema struct",
```
should we use `Schema#sameSchema`?

yyanyy replied:
Have to rebase to incorporate that, but sure!

rdblue commented on core/src/test/java/org/apache/iceberg/TestSchemaID.java:
```diff
@@ -0,0 +1,149 @@
+
+    TestHelpers.assertSameSchemaMap(onlySchemaMap, table.schemas());
+    Assert.assertEquals("Current snapshot's schemaId should be the current",
+        table.schema().schemaId(), (int) table.currentSnapshot().schemaId());
```
This should be okay, but in the future you may want to instead use the `Assert.assertEquals(String, Object, Object)` method. That way, you get the output of the assertion if the schema ID is null, rather than a `NullPointerException`. I think that would be easier to debug.

rdblue commented:
Thanks for fixing this, @yyanyy! I merged it.
